File: RxTools-library/src/main/java/com/vondear/rxtools/view/RxRotateTool.java
package com.vondear.rxtools.view;

import android.graphics.PointF;

/**
 * @author vondear
 */
public class RxRotateTool {
    public static final double CIRCLE_ANGLE = 2 * Math.PI;

    public static double getRotateAngle(PointF p1, PointF p2, PointF mPointCenter) {
        int q1 = getQuadrant(p1, mPointCenter);
        int q2 = getQuadrant(p2, mPointCenter);
        double angle1 = getAngle(p1, mPointCenter);
        double angle2 = getAngle(p2, mPointCenter);
        if (q1 == q2) {
            return angle1 - angle2;
        } else {
            return 0;
        }
    }

    public static double getAngle(PointF p, PointF mPointCenter) {
        float x = p.x - mPointCenter.x;
        float y = mPointCenter.y - p.y;
        double angle = Math.atan(y / x);
        return getNormalizedAngle(angle);
    }

    public static int getQuadrant(PointF p, PointF mPointCenter) {
        float x = p.x;
        float y = p.y;
        if (x > mPointCenter.x) {
            if (y > mPointCenter.y) {
                return 4;
            } else if (y < mPointCenter.y) {
                return 1;
            }
        } else if (x < mPointCenter.x) {
            if (y > mPointCenter.y) {
                return 3;
            } else if (y < mPointCenter.y) {
                return 2;
            }
        }
        return -1;
    }

    public static double getNormalizedAngle(double angle) {
        while (angle < 0)
            angle += CIRCLE_ANGLE;
        return angle % CIRCLE_ANGLE;
    }
}

File: RxUI/src/main/java/com/tamsiree/rxui/view/RxRotateTool.kt
package com.tamsiree.rxui.view

import android.graphics.PointF

/**
 * @author tamsiree
 */
object RxRotateTool {
    const val CIRCLE_ANGLE = 2 * Math.PI
    fun getRotateAngle(p1: PointF, p2: PointF, mPointCenter: PointF): Double {
        val q1 = getQuadrant(p1, mPointCenter)
        val q2 = getQuadrant(p2, mPointCenter)
        val angle1 = getAngle(p1, mPointCenter)
        val angle2 = getAngle(p2, mPointCenter)
        return if (q1 == q2) {
            angle1 - angle2
        } else {
            0.00
        }
    }

    fun getAngle(p: PointF, mPointCenter: PointF): Double {
        val x = p.x - mPointCenter.x
        val y = mPointCenter.y - p.y
        val angle = Math.atan(y / x.toDouble())
        return getNormalizedAngle(angle)
    }

    fun getQuadrant(p: PointF, mPointCenter: PointF): Int {
        val x = p.x
        val y = p.y
        if (x > mPointCenter.x) {
            if (y > mPointCenter.y) {
                return 4
            } else if (y < mPointCenter.y) {
                return 1
            }
        } else if (x < mPointCenter.x) {
            if (y > mPointCenter.y) {
                return 3
            } else if (y < mPointCenter.y) {
                return 2
            }
        }
        return -1
    }

    fun getNormalizedAngle(angle: Double): Double {
        var angle = angle
        while (angle < 0) {
            angle += CIRCLE_ANGLE
        }
        return angle % CIRCLE_ANGLE
    }
}